### PR TITLE
Redo #45 and add a benchmark for unpack

### PR DESCRIPTION
**Pull request by edmundnoble**
Fixes for review of #45.

### CODE REVIEW

File: Codec/Archive/Tar/Types.hs [hunk 1]
```diff
@@ -321,9 +321,9 @@ fromTarPathToPosixPath :: TarPath -> FilePath
 fromTarPathToPosixPath (TarPath namebs prefixbs) = adjustDirectory $ BS.Char8.unpack $
    if BS.null prefixbs
    then namebs
-   else prefixbs <> BS.Char8.pack [FilePath.Posix.pathSeparator] <> namebs
+   else prefixbs <> BS.Char8.singleton FilePath.Posix.pathSeparator <> namebs
   where
-    adjustDirectory | BS.Char8.pack [FilePath.Posix.pathSeparator] `BS.isPrefixOf` namebs
+    adjustDirectory | BS.Char8.singleton FilePath.Posix.pathSeparator `BS.Char8.isSuffixOf` namebs
```
Comment (Bodigrim) [inline Codec/Archive/Tar/Types.hs]: Actually, it does not look to me that you need to adjust directory at all. The previous implementation used to lose the last slash when splitting directories, so it was important to restore it back afterwards. But now you do not trim the last slash from `namebs`, so no reason to add it back. Currently you'll end up with two slashes.

File: Codec/Archive/Tar/Types.hs [hunk 2]
```diff
@@ -342,7 +342,7 @@ fromTarPathToWindowsPath (TarPath namebs prefixbs) = adjustDirectory $
   where
     name   = BS.Char8.unpack namebs
     prefix = BS.Char8.unpack prefixbs
-    adjustDirectory | BS.Char8.pack [FilePath.Posix.pathSeparator] `BS.isPrefixOf` namebs
+    adjustDirectory | BS.Char8.singleton FilePath.Posix.pathSeparator `BS.isPrefixOf` namebs
```
Comment (Bodigrim) [inline Codec/Archive/Tar/Types.hs]: You fixed `isPrefixOf` above, but not here. 

I would use `BS.last` instead of `BS.isSuffixOf`.

**Review by Bodigrim — CHANGES_REQUESTED** [review]
Clearly existing tests do not cover these functions well enough. Please add properties, which check that new functions match their previous definitions.

Also, since the original PR claimed a speed up, could this please be backed by a benchmark?

File: Codec/Archive/Tar/Types.hs (hunk 2 above)
Comment (edmundnoble) [reply Codec/Archive/Tar/Types.hs]: Is it also necessary to check that `namebs` is non-empty?

Comment (Bodigrim) [reply Codec/Archive/Tar/Types.hs]: Yes.

Comment (edmundnoble) [reply Codec/Archive/Tar/Types.hs]: That was why I used `isPrefixOf`. I'll change it.

**Comment by edmundnoble** [comment]
I have dredged up the old project that was benchmarked with #45. It uses `Tar.unpack`. I can't see right now, looking at #45, why it would improve (time) performance of `unpack`, because `entryPath` uses `fromTarPath`, not `fromTarToPosixPath`. In fact, I don't believe that my project actually benchmarked time performance *at all*, I think I extrapolated from the .prof file. Redoing that "benchmark", I see `fromTarPath` implicated instead. I don't see why that would've changed. I'm confused.

Here, benchmarking `unpack` with a version of `fromTarPath` which just uses the "optimized" `fromTarToPosixPath`, I see a reduction in allocations by one sixth, no reduction in live set size, and no reduction in time taken. To me this conclusively proves that #45 does not substantially improve performance.

Thus I've replaced the contents of this PR with a benchmark for `unpack` and a revert of the previous commit. Sorry about that.

**Comment by Bodigrim** [comment]
> I see a reduction in allocations by one sixth

Reduction in allocations is nevertheless a very worthy goal, even if it does not lead to an immediate speed up. And I think it is a proper thing for `fromTarPathToPosixPath` to rely that `prefixbs` and `namebs` are POSIX paths already and do not split/join them again. So in general I'm in favor of your patch.

We just need to grow more confidence that the behaviour did not change, especially on non-Ubuntu platforms. As a first step, one could submit a multiplatform CI workflow for the latest GHC, borrowing relevant parts from https://github.com/haskell/bytestring/blob/master/.github/workflows/ci.yml

**Comment by edmundnoble** [comment]
Does the latest form of the change satisfy you that there are no changes on non-POSIX platforms? I can take a look at CI too, but it may take longer.

**Comment by edmundnoble** [comment]
CI is at #62.

**Comment by Bodigrim** [comment]
Superseded by ffc123f, thanks for trailblazing the path!

**Pull request closed without merging**